### PR TITLE
HBASE-23787: TestSyncTimeRangeTracker fails quite easily and allocate…

### PR DESCRIPTION
**Pull request by markrmiller**
…s a very expensive array.

Here is a first pass, says it can't merge for some reason so I will be back for draft 2.

### CODE REVIEW

**Comment by saintstack** [comment]
Your branch up-to-date Mr. Miller?

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java [hunk 1]
```diff
@@ -84,23 +86,23 @@ public void run() {
...
   }
 
-  static class RandomTestData {
-    private long[] keys = new long[NUM_KEYS];
-    private long min = Long.MAX_VALUE;
-    private long max = 0;
+ static class RandomTestData {
+    private final AtomicLongArray keys = new AtomicLongArray(NUM_KEYS);
+    private long min = Long.MAX_VALUE; // effectively final
+    private long max = 0; // effectively final
```
Comment (saintstack) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: Don't want to make them final?

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java [hunk 2]
```diff
@@ -84,23 +86,23 @@ public void run() {
...
 
     public RandomTestData() {
       if (ThreadLocalRandom.current().nextInt(NUM_OF_THREADS) % 2 == 0) {
         for (int i = 0; i < NUM_KEYS; i++) {
-          keys[i] = i + ThreadLocalRandom.current().nextLong(NUM_OF_THREADS);
-          if (keys[i] < min) min = keys[i];
-          if (keys[i] > max) max = keys[i];
+          keys.set(i, i + ThreadLocalRandom.current().nextLong(NUM_OF_THREADS));
+          if (keys.get(i) < min) min = keys.get(i);
```
Comment (saintstack) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: See your checkstyle. Needs parents around the 'min = key[i];'

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java [hunk 3]
```diff
@@ -84,23 +86,23 @@ public void run() {
     assertTrue(trr.getMin() == 0);
   }
 
-  static class RandomTestData {
-    private long[] keys = new long[NUM_KEYS];
-    private long min = Long.MAX_VALUE;
-    private long max = 0;
+ static class RandomTestData {
+    private final AtomicLongArray keys = new AtomicLongArray(NUM_KEYS);
```
Comment (saintstack) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: This is the 'fix'? Using atomic array. Sounds good.

Comment (Apache9) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: But this array will only be accessed by one thread? Why do we need to use AtomicLongArray?

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java [hunk 4]
```diff
@@ -34,7 +36,7 @@
   public static final HBaseClassTestRule CLASS_RULE =
       HBaseClassTestRule.forClass(TestSyncTimeRangeTracker.class);
 
-  private static final int NUM_KEYS = 10000000;
+  private static final int NUM_KEYS = 1000000;
```
Comment (Apache9) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: So this is the actual fix right?

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java (hunk 1 above)
Comment (busbey) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: they get assigned like right below here.

**Comment by saintstack** [comment]
@markrmiller See comments above?

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java (hunk 3 above)
Comment (markrmiller) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: Basically I ran this test a lot before and after these changes to try and confirm a fix and this is what seemed to work.

Technically, even if a single thread accesses a data structure, if it is a different thread than the one that created it, you still need to publish objects correctly for multiple threads - concurrency is not required. So they need a mem barrier or to be final or to be effectively final or a 32-bit primitive(I think?).

Comment (markrmiller) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: Yeah, so AtomicLongArray to ensure we see an up to date value.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java (hunk 4 above)
Comment (markrmiller) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: This is kind of separate. I noticed this huge array took a lot of RAM and it did not seem necessary.

I also noticed that the array is so large that the threads working on it don't overlap much as is.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java (hunk 1 above)
Comment (markrmiller) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSyncTimeRangeTracker.java]: Right, would be ideal to make them final, settling for marking them effectively final.

**Comment by markrmiller** [comment]
Sorry, been chasing a local clean test run.

The AtomicLongArray was need for these tests to pass more regularly for me to ensure we always look at an up to date value.

The memory issue seemed wasteful and a bit off given the goals of the test and also counter productive to running test like this in parallel.

I've updated my pr.

**Comment by saintstack** [comment]
No response on above. Closing this out.

**Pull request closed without merging**